### ISSUE TITLE
[Bug?] subtle influence of/by preact "render": wrong effect function invoked

### ISSUE THREAD

**Issue by rozek** [melnikov-s/preact-observables#2]
First of all: thank you very much for this module!

But I have the impression, that it does not work properly - because of a subtle (but really nasty) influence of (or by) the preact "render" function

Consider the following code (do not ask for any deeper sense, I have tried to extract the essence of a much larger application with dynamic plugins):

```
  import { render, html, effect }   from 'preact-with-htm'
  import { observable, Observable } from 'preact-observables'

  const global = (new Function('return this'))()

  global.Store = observable({
    Value:0
  })

  class Hugo extends Observable {
    _Store = Object.create(null)
  }

  const hugo = new Hugo()

  Object.defineProperty(hugo, 'Value', {
    configurable:true, enumerable:true,
    get () { return this._Store.Value},
    set (newValue) { this._Store.Value = newValue; this.render() }
  })

  hugo.render = function () {       // defined before any effects with "render"!
    render(html`Value is now ${hugo.Value}`,document.body)
  }
  hugo.render()

  effect(() => {
    console.log('deriving "hugo.Value" from "global.Value"',global.Store.Value)
    hugo.Value = global.Store.Value
  })
  effect(() => {
    console.log('updating "global.Value" from "hugo.Value"',hugo.Value)
    global.Store.Value = hugo.Value
  })

  console.log('\n\nsetting "hugo.Value" to 1')
  hugo.Value = 1

  console.log('\n\nsetting "global.Value" to 2')
  global.Store.Value = 2
```

which produces the following output:

```
deriving "hugo.Value" from "global.Value" 0
updating "global.Value" from "hugo.Value" 0

setting "hugo.Value" to 1
deriving "hugo.Value" from "global.Value" 0
updating "global.Value" from "hugo.Value" 0

setting "global.Value" to 2
deriving "hugo.Value" from "global.Value" 2
updating "global.Value" from "hugo.Value" 2
```

As you can see, after setting `hugo.Value` to `1`, the wrong "effect" function is called which sets `hugo.Value` back to 0 (the current `global.Store.Value`)

Now, disable the `render` invocation in the value setter and run the same code again. It will now output:

```
deriving "hugo.Value" from "global.Value" 0
updating "global.Value" from "hugo.Value" 0

setting "hugo.Value" to 1
updating "global.Value" from "hugo.Value" 1
deriving "hugo.Value" from "global.Value" 1

setting "global.Value" to 2
deriving "hugo.Value" from "global.Value" 2
updating "global.Value" from "hugo.Value" 2
```

which is correct.

Even worse: if you re-enable the `render` invocation but simply swap the two `effect` calls (like so

```
  effect(() => {
    console.log('updating "global.Value" from "hugo.Value"',hugo.Value)
    global.Store.Value = hugo.Value
  })
  effect(() => {
    console.log('deriving "hugo.Value" from "global.Value"',global.Store.Value)
    hugo.Value = global.Store.Value
  })
```

) and run the code again, its output will be correct again:

```
updating "global.Value" from "hugo.Value" undefined
 deriving "hugo.Value" from "global.Value" undefined

setting "hugo.Value" to 1
updating "global.Value" from "hugo.Value" 1
deriving "hugo.Value" from "global.Value" 1

setting "global.Value" to 2
deriving "hugo.Value" from "global.Value" 2
updating "global.Value" from "hugo.Value" 2
```

This is a really severe problem which makes my code (using preact-observables) practically unusable as I cannot predict its behaviour.

Do you have any idea what the reason for this strange (race condition?) could be?

**Comment by rozek** [comment]
~~For those who may come across the same problem: it might be that a workaround could be to change `this.render` to `setImmediate(() => this.render())` or `setTimeout(() => this.render(),0)`~~

~~However, this would just be a really ugly workaround as it does not solve the underlying problem but only this single "manifestation" of it - and may cause other unwanted side-effects.~~

**Comment by melnikov-s** [comment]
Thanks for the issue! I should have time to take a look at it this week, do you happen to have a codesandbox or stackblitz link handy with the reproduction?

**Comment by melnikov-s** [comment]
Thanks for raising this. I had a chance to look into it and few things worth mentioning

 This has nothing to do with the `render` call but rather the `hugo.Value` getter. If you remove everything but that in `render` it will still trigger this issue.

I'm not sure why you're rendering in a setter, render should be an effect.  ~Either way this does look like a legitimate bug though it's not a bug in `preact-observables` I can reproduce it with just `preact/signals-core`@1.5.0 . You might want to raise this issue there.~ It's not a bug it seems to be a "quirk" of the dependency graph , since you're observing a computed in there and it depends on the observable that ran in the second effect the order of effects that are ran gets changed depending whether or not the  computed is present in the first effect. This exact same behavior happens with other reactive libraries like `mobx` .  

Since this is a contrived example I can't really say what you're doing wrong but you're likely using an anti-pattern.

**Comment by melnikov-s** [comment]
Here's a simplified reproduction with `preact/signals-core`

```javascript
import { effect, signal, computed } from '@preact/signals-core';

const a = signal(0);
const b = signal(0);
const comp = computed(() => b.value);

effect(() => {
  console.log('deriving "b" from "a"', a.value);
  b.value = a.value;
  comp.value
});

effect(() => {
  console.log('updating "a" from "b"', b.value);
  a.value = comp.value;
});

console.log('\n\nsetting "b" to 1');
b.value = 1;
console.log('\n\nsetting "a" to 2');
a.value = 2;
```

uncommenting `comp.value` in the first `effect` produces the "correct" results

**Comment by rozek** [comment]
Thank you very much for your effort - particularly for the simplified version of my example!

One of my problems (in the real application) was that the "effect"s in the above example actually are "plugins" from other users which I do not have under my control - although I would like to assert that they behave "as expected".

However, what I indeed have under my control is `comp.value` (or its counterpart in the real code) - and I surrounded that with a call of `untracked`. Since then, everything works as intended - although the solution still smells a bit like a hack or ugly workaround...

So, again, thank you very much for your help!
